app/controllers/application_controller.rb

```diff
@@ -10,24 +10,20 @@ class ApplicationController < ActionController::Base
   protected
   # We are overriding this method to account for permission trees. See
   # TreePolicy
-  def authorize(record, query=nil)
-    # use the action as a default permission
-    query ||= (params[:action].to_s + "?").to_sym
-
-    pol = policy(record)
-    # if an instance of TreePolicy, convert the query into the base/"real"
-    # permissions (see TreePolicy#flatten_tree)
-    if pol.respond_to?(:flatten_tree)
-      queries = pol.flatten_tree(query)
-    else
-      queries = [query]
-    end
+  def authorize(record, query=nil, user=nil)
+    user ||= @current_user
+    policy = Pundit.policy(user, record)
 
-    queries.each do |q|
-      super(record, q)  # have Pundit process each permission
+    # use the action as a default permission
+    query ||= ("can_" + params[:action].to_s + "!").to_sym
+    unless policy.public_send(query)
```

Comment:
I was under the assumption that `params[:action]` was a whitelist already?

All queries should be coming from code rather than a user. If we make a mistake and allow user input to get in, however, that user would only have access to the policy object, which is intentionally limited.